request_test.go

```diff
@@ -1601,3 +1603,85 @@ func TestRequestErrorConflictMultipart(t *testing.T) {
 	req3.WithFileBytes("a", "a", []byte("a"))
 	req3.chain.assertFailed(t)
 }
+
+func TestRequestRedirect(t *testing.T) {
+	reporter := newMockReporter(t)
+
+	t.Run("dont follow redirects policy", func(t *testing.T) {
+		client := &http.Client{
+			Transport: mockTransportWithNoRedirect{t: t},
+		}
+
+		config := Config{
+			Client:   client,
+			Reporter: reporter,
+		}
+
+		req := NewRequest(config, "METHOD", "url").
+			WithHeader("Location", "/redirect").
+			WithRedirectPolicy(DontFollowRedirects)
+		req.chain.assertOK(t)
+
+		resp := req.Expect().
+			Status(http.StatusPermanentRedirect)
+		resp.chain.assertOK(t)
+
+		httpClient, _ := req.config.Client.(*http.Client)
+		assert.Equal(t, http.ErrUseLastResponse, httpClient.CheckRedirect(req.httpReq, nil))
+
+		assert.Nil(t, req.httpReq.GetBody)
+	})
+
+	t.Run("follow all redirect", func(t *testing.T) {
+		client := &http.Client{
+			Transport: mockTransportWithNoRedirect{t: t},
+		}
+
+		config := Config{
+			Client:   client,
+			Reporter: reporter,
+		}
+
+		req := NewRequest(config, "METHOD", "url").
+			WithHeader("Location", "/redirect").
+			WithRedirectPolicy(FollowAllRedirects).
+			WithMaxRedirects(0).
+			WithText("test body")
+		req.chain.assertOK(t)
+
+		resp := req.Expect().
+			Status(http.StatusPermanentRedirect)
+		resp.chain.assertOK(t)
+
+		httpClient, _ := req.config.Client.(*http.Client)
+		assert.Equal(t, nil, httpClient.CheckRedirect(req.httpReq, make([]*http.Request, 0)))
+		assert.EqualError(t,
+			httpClient.CheckRedirect(req.httpReq, make([]*http.Request, 1)),
+			fmt.Sprintf("stopped after %d redirects", req.maxRedirects),
+		)
+
+		rd, err := req.httpReq.GetBody()
+		assert.NoError(t, err)
+		b, err := ioutil.ReadAll(rd)
+		assert.NoError(t, err)
+		assert.Equal(t, "test body", string(b))
+	})
+}
+
+type mockTransportWithNoRedirect struct {
+	t         *testing.T
+	tripCount uint8
+}
+
+func (mt mockTransportWithNoRedirect) RoundTrip(origReq *http.Request) (
+	*http.Response, error,
+) {
+	mt.tripCount++
+
+	assert.Equal(mt.t, mt.tripCount, uint8(1))
+
+	res := httptest.NewRecorder()
+	res.Result().StatusCode = http.StatusPermanentRedirect
+
+	return res.Result(), nil
+}
```

Comment:
In the second test, let's also check here that origReq.Body is set and we can read from it and it gives us correct value ("test body").